---
title: "Understanding Sequencing Reads"
author: "Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: html_document
---


# Hands-on Practice

## Overview

- Understanding the contents of a fastq file
- How fastq files can be accessed in Bioconductor
    + (should you ever need to)

Two example fastq files have been provided for you in the folder `/home/participant/Course_Materials/Day1/`. As we discussed, they can be viewed without any specialist software. If you are curious how these fastq files were generated, you can see the [Appendix](#appendix) for details.

We can open a ***Terminal*** window and print the first few lines using `head`

```{}
cd /home/participant/Course_Materials/Day1/
```


```{}
head -n 12 sample.fq1
```

```{r, echo=FALSE,comment=NA}
cat(system("head -n12 sample.fq1",intern=TRUE),sep="\n")
```

```{}
head -n 12 sample.fq2
```


```{r, echo=FALSE,comment=NA}
cat(system("head -n12 sample.fq2",intern=TRUE),sep="\n")
```

The unix command `wc` can count the number of lines in a file with the option `-l`

```{}
wc -l sample1.fq1
```

```{r echo=FALSE,comment=NA}
cat(system("wc -l sample.fq1",intern=TRUE),sep="\n")
```



We will now use the `ShortRead` Bioconductor package to understand these files a bit more. Please note that we are using this package for exploration and to demonstrate what the data look like. We probably wouldn't want to perform these tasks in a production-level analysis, and there are far superior tools for visualising fastq files (as we will see later).



```{r echo=FALSE,message=FALSE}
library(ShortRead)
```

Before you begin, you'll need to make sure that your *working directory* is set to `/home/participant/Course_Materials/Day1/`. The easiest way to do this is to navigate the RStudio menus

***Session*** -> ***Set Working Directory*** -> ***Choose Directory***

******
### Exercise

Import the example `fastq` file using the `ShortRead` package.

- How many reads are found?
- How long is each sequenced read?

```{r message=FALSE, results='hide'}
library(ShortRead)
fq <- readFastq("sample.fq1")
fq
```

******

The object we have just created is typical of how data are presented in Bioconductor packages; typing the name of the object `fq` prints a summary of the contents to the screen, and we have to use a series of functions to extract the contents. Such functions are referred to as *accessor functions*. 

One such function is called `sread` and allows us to access the sequencing reads. Conveniently, the function prints a summary to the screen so that only the first five and last five reads in the file are displayed. The type of this object is a `DNAStringSet`, which is a very efficient way of storing DNA sequences in Bioconductor. The `Biostrings` package has lots of useful tools to manipulate these objects. We can save the result of `sread` to a variable for futher processing.


```{r}
sread(fq)
myreads <- sread(fq)
```

Even though the data they represent are much more complex, the way in which we subset the data should be familiar to us. This means there are no new rules to learn when it comes to accessing the data.

Try the following:-

```{r}
myreads[1:5]
```

We can treat the `myreads` object as a vector and use the `[]` in the normal fashion. However, it might be tempting to think that two indexes could be used to subset both the read index ,and the position in the read.

```{r eval=FALSE}
myreads[1:5,1:10]
```

Alas, the object is not designed to work in this way and we get an error message.

```{r echo=FALSE}
t <- try(myreads[1:5,1:3],silent=TRUE)
cat(t)
```

If we want to chop-up the strings in this manner, we can use the `subseq` function. 

******

### Exercise
- Use `subseq` to print the first five basees of the sequenced reads

```{r}
firstBases <- subseq(myreads,1,5)
firstBases
```


- How many reads start with five A's?
    + HINT: we can test for equality using `==` which gives a logical (`TRUE` or `FALSE`)
    + HINT: `table` can be used to get counts of how many different values occur in a vector
- What is the most-common five bases at the start of the reads
    + HINT: `sort` will produce a sorted version of a vector

******



As we mentioned, `Biostrings` contains many useful functions for dealing with DNA sequences. Many man-hours have been spent trying to re-implement functions that are already provided in this package; often with less-efficient results! You can consult the help page for `Biostrings` (which can be found via the Packages tab in the bottom-right panel of RStudio) for a full list. 


The *"base"* distribution of R offers many string manipulation functions. One is `substr` that will perform the trimming of characters in the same manner. However, `substr` is a lot slower as it has not been optimised for large datasets of biological sequences. 

Moral: always check if there is a function in `Biostrings` that you can use rather than a base function.
```{r}
system.time(subseq(myreads, 1,5))
system.time(substr(myreads, 1,5))
```


Another useful function is `letterFrequency`, which seeks to tabulate the number of letters in each sequence in the form of a matrix. 

```{r}
af <- letterFrequency(sread(fq), letters=c("A","C","G","T","N"))
head(af)
```

This matrix can be accessed in the usual manner. For instance, we could get the number of `A` bases for each read, as a vector, and produce a histogram with the following two lines:-

```{r}
aS <- af[,1]
hist(aS)
```

******

### Exercise

- What do you think the N column could represent?
- Which read has the greatest number of N bases?
- Create a vector to represent the "GC" content of each read
- What is the mean number of GC bases?
- Visualise the GC distribution using a histogram
- What reads have more than 60 GC bases?

The following R output is for your reference

```{r echo=FALSE}
myreads[which.max(af[,5])]
gc <- af[,"C"] + af[,"G"]
hist(gc)
myreads[gc > 60]
badGCReads <- which(gc > 60)
``` 

******

Note that at this point we have not performed any manipulation on the reads themselves. If we wanted to make persistent changes to the object, such as creating a subset, we would have to overwrite the `myreads` object (not recommended!) or create a new object

```{r}
myreads
```


We can remove reads using the negative index approach

```{r eval=FALSE}
goodReads <- myreads[-badGCReads]
```


We can also look in more detail at the number of bases that were called during each round of sequencing

```{r}
abc <- alphabetByCycle(myreads,alphabet = c("A","C","G","T","N"))
head(abc)[,1:5]
```

We can visualise the number of N bases, and how they increase as sequencing progresses. Observing more N bases towards the end of the sequencing run is common, but not always a cause for alarm.

```{r}
plot(1:68, abc[5,1:68],type="n")
text(1:68, abc[5,1:68], label="N")
```

******

### Exercise


Can you make the following plot to illustrate the number of each DNA base observed at each position?

```{r echo=FALSE}
plot(abc[1,],type="n",ylim=c(0,2000000))
text(abc[1,1:68],label="A",col="red")
text(abc[2,1:68],label="C",col="orange")
text(abc[3,1:68],label="G",col="blue")
text(abc[4,1:68],label="T",col="green")
```

******


Also in the fastq file is the *quality* associated with each base call. This is especially important when it comes to calling SNVs, as we need to be sure of the bases we observe. The `quality` function is used to access these data.

```{r}
quality(fq)
myquals <- quality(fq)
```

As we noted before, a variey of [different schemes can be used to encode the qualities](https://en.wikipedia.org/wiki/FASTQ_format#Encoding). Fortunately, with the help of the `encoding` function, we can make a guess about what scale is being used in our file

```{r}
encoding(quality(fq))
```

******

### Exercise

- What can you say about the quality of base calls in the first five reads?
- What reads seem to have lower quality? 
    + Is it a particular part of the read that is low quality? 
    + Or the whole read?

******



Quality scores can often be used to motivate the need for trimming, or to use an aligner that can incorporate quality information. We can use the following two commands to obtain a more-familiar matrix of numeric values to represent quality

```{r}
qm <- as(quality(fq), "matrix")
head(qm)
```


To look at global trends, it is useful to look at the mean quality at each base. The `colMeans` function in R provides a quick way of being able to compute the average (mean) of each column in a *matrix*. 

```{r}
qualmeans <- colMeans(qm)
plot(qualmeans)
```


Hopefully this section was useful in familiarising yourself with the contents of a `fastq` file and give ideas about how R and Bioconductor can interface with sequencing data. However, in practice we would not normally look into fastq files in such depth in R. 
