---
title: "Understanding Sequencing Reads"
author: "Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: html_document
---


# Hands-on Practice

## Overview

- Understanding the contents of a `.fastq` file
- How `.fastq` files can be accessed in Bioconductor
    + (should you ever need to
- What information in a `.fastq` file can inform us about quality
- How complex data structures can be assessed in Bioconductor

Two example `.fastq` files have been provided for you in the folder `/home/participant/Course_Materials/Day1/`. As we discussed, they can be viewed without any specialist software. If you are curious how these fastq files were generated, you can see the [Appendix](#appendix) for details.

We can launch the ***CRUK Docker*** from the Desktop and navigate to the directory containing the files

```{}
cd /home/participant/Course_Materials/Day1/
```

Then print the first few lines using the standard unix command `head`

```{}
head -n 12 sample.fq1
```

```{r, echo=FALSE,comment=NA}
cat(system("head -n12 sample.fq1",intern=TRUE),sep="\n")
```

```{}
head -n 12 sample.fq2
```


```{r, echo=FALSE,comment=NA}
cat(system("head -n12 sample.fq2",intern=TRUE),sep="\n")
```

The unix command `wc` can count the number of lines in a file with the option `-l`

```{}
wc -l sample.fq1
```

```{r echo=FALSE,comment=NA}
cat(system("wc -l sample.fq1",intern=TRUE),sep="\n")
```



We will now use the `ShortRead` Bioconductor package to understand these files a bit more. Please note that we are using this package for exploration and to demonstrate what the data look like. We probably wouldn't want to perform these tasks in a production-level analysis, and there are far superior tools for visualising fastq files (as we will see later).


## Getting started

Load RStudio....

### Working directory

***Session*** -> ***Set Working Directory*** -> ***Choose Directory***

```{}
/home/participant/Course_Materials/Day1/
```

### Template

```{}
/home/participant/Course_Materials/Day1/Session2-template.Rmd
```



******
******
******
### Exercise

Import the example `fastq` file using the `ShortRead` package.

- How many reads are found?
- How long is each sequenced read?



```{r message=FALSE, results='hide'}
library(ShortRead)
fq <- readFastq("sample.fq1")
fq
```

******
******
******

The object we have just created is typical of how data are presented in Bioconductor packages; typing the name of the object `fq` prints a summary of the contents to the screen, and we have to use a series of functions to extract the contents. Such functions are referred to as *accessor functions*. 

One such function is called `sread` and allows us to access the sequencing reads. Conveniently, the function prints a summary to the screen so that only the first five and last five reads in the file are displayed. The type of this object is a `DNAStringSet`, which is a very efficient way of storing DNA sequences in Bioconductor. The `Biostrings` package has lots of useful tools to manipulate these objects. We can save the result of `sread` to a variable for futher processing.


```{r}
sread(fq)
myreads <- sread(fq)
```

Even though the data they represent are much more complex, the way in which we subset the data should be familiar to us. This means there are no new rules to learn when it comes to accessing the data.

Try the following:-

```{r}
myreads[1:5]
```

We can treat the `myreads` object as a vector and use the `[]` in the normal fashion. However, it might be tempting to think that two indexes could be used to subset both the read index ,and the position in the read.

```{r eval=FALSE}
myreads[1:5,1:10]
```

Alas, the object is not designed to work in this way and we get an error message.

```{r echo=FALSE}
t <- try(myreads[1:5,1:3],silent=TRUE)
cat(t)
```

If we want to chop-up the strings in this manner, we can use the `subseq` function. The is a specialised function that is much-more efficient that the general `substr` function in base R.

******
******
******

### Exercise
- Use `subseq` to print the first five bases of the sequenced reads

```{r}
firstBases <- subseq(myreads,1,5)
firstBases
```


- How many reads start with five A's?
    + HINT: we can test for equality using `==` which gives a logical (`TRUE` or `FALSE`)
    + HINT: `table` can be used to get counts of how many different values occur in a vector
- What is the most-common five bases at the start of the reads
    + HINT: `sort` will produce a sorted version of a vector

******
******
******


Another useful function is `letterFrequency`, which seeks to tabulate the number of letters in each sequence in the form of a matrix. 

```{r}
af <- letterFrequency(sread(fq), letters=c("A","C","G","T","N"))
head(af)
```

This matrix can be accessed in the usual manner. For instance, we could get the number of `A` bases for each read, as a vector, and produce a histogram with the following two lines:-

```{r}
aS <- af[,1]
hist(aS)
```

******
******
******
### Exercise

- Create a vector to represent the "GC" content of each read
- What is the mean number of GC bases?
- Visualise the GC distribution using a histogram
- What reads have more than 60 GC bases?

The following R output is for your reference

```{r echo=FALSE}
myreads[which.max(af[,5])]
gc <- af[,"C"] + af[,"G"]
hist(gc)
myreads[gc > 60]
badGCReads <- which(gc > 60)
``` 

******
******
******

Note that at this point we have not performed any manipulation on the reads themselves. If we wanted to make persistent changes to the object, such as creating a subset, we would have to overwrite the `myreads` object (not recommended!) or create a new object

```{r}
myreads
```



We can also look in more detail at the number of bases that were called during each round of sequencing

```{r}
abc <- alphabetByCycle(myreads,alphabet = c("A","C","G","T","N"))
head(abc)[,1:5]
```


The following plot can be used to illustrate the number of each DNA base observed at each position

```{r}
plot(abc[1,],type="n",ylim=c(0,2000000))
text(abc[1,1:68],label="A",col="red")
text(abc[2,1:68],label="C",col="orange")
text(abc[3,1:68],label="G",col="blue")
text(abc[4,1:68],label="T",col="green")
```

Also in the `.fastq` file is the *quality* associated with each base call. This is especially important when it comes to calling SNVs, as we need to be sure of the bases we observe. The `quality` function is used to access these data.

```{r}
quality(fq)
myquals <- quality(fq)
```

As we noted before, a variey of [different schemes can be used to encode the qualities](https://en.wikipedia.org/wiki/FASTQ_format#Encoding). Fortunately, with the help of the `encoding` function, we can make a guess about what scale is being used in our file

```{r}
encoding(quality(fq))
```

******
******
******

### Exercise

- What can you say about the quality of base calls in the first five reads?
- What reads seem to have lower quality? 
    + Is it a particular part of the read that is low quality? 
    + Or the whole read?

******
******
******


Quality scores can often be used to motivate the need for trimming, or to use an aligner that can incorporate quality information. We can use the following two commands to obtain a more-familiar matrix of numeric values to represent quality

```{r}
qm <- as(quality(fq), "matrix")
head(qm)
```


To look at global trends, it is useful to look at the mean quality at each base. The `colMeans` function in R provides a quick way of being able to compute the average (mean) of each column in a *matrix*. 

```{r}
qualmeans <- colMeans(qm)
plot(qualmeans)
```


Hopefully this section was useful in familiarising yourself with the contents of a `fastq` file and give ideas about how R and Bioconductor can interface with sequencing data. However, in practice we would not normally look into fastq files in such depth in R. 



# <a name="appendix"></a> Appendix 

For your reference, here is how the example files were created

First, we download an example bam file from the 1000 genomes project

```{}
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/other_exome_alignments/NA06984/exome_alignment/NA06984.mapped.illumina.mosaik.CEU.exome.20111114.bam
```

Then the file is *downsampled* to give a much reduced set of reads. The original file can be removed
```{}
java -jar $PICARD DownsampleSam I=NA06984.mapped.illumina.mosaik.CEU.exome.20111114.bam O=random.bam P=0.1 VALIDATION_STRINGENCY=SILENT

rm NA06984.mapped.illumina.mosaik.CEU.exome.20111114.bam
```

For convenience, we filter the file to keep only reads that are properly paired. Then picard is used again to extract the fastq data from the file. As the original alignments were a mix of 68 and 76 bases, we trim all the reads to 68 bases to processing easier.

```{}
samtools view -f 0x02 -b random.bam > paired.bam 

java -jar $PICARD SamToFastq I=paired.bam F=sample.fq1 VALIDATION_STRINGENCY=SILENT F2=sample.fq2 R1_MAX_BASES=68 R2_MAX_BASES=68
```


